### PR TITLE
Implemented support for Advanced joins in BigQuery Pushdown

### PR DESCRIPTION
**Pull request by fernst**
Added further tests and code improvements.

Refactored SQL Join builder to take Join Definition as input and expose simple method to get join SQL.

### CODE REVIEW

File: src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQuerySQLEngine.java [hunk 1]
```diff
@@ -171,20 +173,33 @@ public boolean exists(String datasetName) throws SQLEngineException {
...
   @Override
   public boolean canJoin(SQLJoinDefinition sqlJoinDefinition) {
-    // TODO: remove when expression joins are supported.
-    if (sqlJoinDefinition.getJoinDefinition().getCondition().getOp() != JoinCondition.Op.KEY_EQUALITY) {
-      return false;
-    }
 
-    //Ensure none of the schemas contains unsupported types.
-    for (JoinStage inputStage : sqlJoinDefinition.getJoinDefinition().getStages()) {
+    List<String> issues = new LinkedList<>();
```
Comment (tivv) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQuerySQLEngine.java]: How is this issues list used? I see you add data to it, but not retrieve. Also please use ArrayList. LinkedList makes sense only in very edge cases. In most cases ArrayList is faster and use less memory.

File: src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQueryJoinDataset.java [hunk 2]
```diff
@@ -72,7 +69,7 @@ private BigQueryJoinDataset(String datasetName,
     this.bqDataset = bqDataset;
     this.bqTable = bqTable;
     this.jobId = jobId;
-    this.helper = new BigQuerySQLBuilder(this.project, this.bqDataset, stageToTableNameMap);
+    this.helper = new BigQuerySQLBuilder(this.joinDefinition, this.project, this.bqDataset, stageToTableNameMap);
```
Comment (tivv) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQueryJoinDataset.java]: please name this `queryBuilder` or similar. It's better to have specific names

File: src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQuerySQLEngine.java [hunk 3]
```diff
@@ -171,20 +173,33 @@ public boolean exists(String datasetName) throws SQLEngineException {
...
+
+    JoinDefinition joinDefinition = sqlJoinDefinition.getJoinDefinition();
+
+    // Ensure none of the input schemas contains unsupported types or invalid stage names.
+    for (JoinStage inputStage : joinDefinition.getStages()) {
       if (!supportsSchema(inputStage.getSchema())) {
-        return false;
+        issues.add(String.format("Input schema from stage '%s' has one of more unsupported field types.",
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQuerySQLEngine.java]: of -> or

Comment (albertshau) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQuerySQLEngine.java]: it would be more useful if the message contained which field is unsupported

File: src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQuerySQLEngine.java [hunk 4]
```diff
@@ -171,20 +173,33 @@ public boolean exists(String datasetName) throws SQLEngineException {
...
+                                 inputStage.getStageName()));
       }
     }
 
-    if (!supportsSchema(sqlJoinDefinition.getJoinDefinition().getOutputSchema())) {
-      return false;
+    // Ensure the output schema doesn't contain unsupported types
+    if (!supportsSchema(joinDefinition.getOutputSchema())) {
+      issues.add("Output Schema has one of more unsupported field types.");
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQuerySQLEngine.java]: similar comment here, message should include which field(s) is unsupported

File: src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQuerySQLEngine.java [hunk 5]
```diff
@@ -349,4 +365,18 @@ protected boolean supportsSchema(@Nullable Schema schema) {
...
 
+  /**
+   * Ensure the Stage name is valid for execution in BQ pushdown.
+   *
+   * Due to differences in character escaping rules in Spark and BigQuery, identifiers that are accepted in Spark
+   * might not be valid in BigQuery. Due to this limitation, we don't support stage names or aliases containing
+   * backslash \ or backtick ` characters at this time.
+   *
+   * @param identifier stage name or alias to validate
+   * @return wether this stage name is valid for BQ Pushdown.
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQuerySQLEngine.java]: typo: whether

File: src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/builder/BigQuerySQLBuilder.java [hunk 6]
```diff
@@ -24,33 +24,90 @@
...
+ * <p>
+ * Many methods in this class have proteced visibility for the purposes of testing.
  */
 public class BigQuerySQLBuilder {
+
+  /**
+   * Allows us to find a table name in the user supplied advanced SQL Join expression.
+   * Explanation of groups:
+   * 1. Matches on 3 possibilities:
+   *   a. The beggining of the string,
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/builder/BigQuerySQLBuilder.java]: typo: beginning

File: src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/builder/BigQuerySQLBuilder.java [hunk 7]
```diff
@@ -24,33 +24,90 @@
...
+   *   b. A non-word character (like a parenthesys or symbol)
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/builder/BigQuerySQLBuilder.java]: typo: parantheses

File: src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/builder/BigQuerySQLBuilder.java [hunk 8]
```diff
@@ -24,33 +24,90 @@
...
+   * <p>
+   * This can match expressions such as:
+   * table.column
+   * `table`.column
+   * table .column
+   * `table` .column
+   * table    .  column
+   * `table`    .  column
+   */
+  private static final String TABLE_NAME_PATTERN = "(^|\\W|^\\.\\s*)(`?)(%s)(`?)(\\s*)(\\.)";
```
Comment (tivv) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/builder/BigQuerySQLBuilder.java]: `^\\.\\s*` does not mean "not a dot". It means start of line followed by dot. "^" would mean not only inside square brackets.
Please add separate unit tests to ensure the pattern does what's expected

File: src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/builder/BigQuerySQLBuilder.java [hunk 9]
```diff
@@ -106,28 +189,127 @@ protected void appendFieldEqualityClause(JoinDefinition joinDefinition) {
...
+
+    String replacement1 = String.format(TABLE_NAME_PATTERN, Pattern.quote(leftIdentifier));
+    String replacement2 = String.format(TABLE_NAME_PATTERN, Pattern.quote(rightIdentifier));
+    String replacement3 = String.format(TABLE_NAME_PATTERN, Pattern.quote(temp));
+
+
+    // Replace the left idenfier with the left alias, and the right identifier with the right alias.
+    // Note that Left is transformer to the temporary placeholder before flipping to the right value.
+    String replaced = expression
+      .replaceAll(replacement1, "$1" + Matcher.quoteReplacement(temp) + "$5$6")
```
Comment (tivv) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/builder/BigQuerySQLBuilder.java]: nit: consider using named groups, it would be a bit more readable. Also you can join groups 5 and 6.

File: src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/builder/BigQuerySQLBuilder.java [hunk 10]
```diff
@@ -106,28 +189,127 @@ protected void appendFieldEqualityClause(JoinDefinition joinDefinition) {
...
+   * Executes replace operation to replace table names in the supplied string for BigQuery-safe table aliases.
+   *
+   * @param expression       expression to manipulate
+   * @param leftIdentifier   Left Stage name
+   * @param leftReplacement  Left Stage replacement alias
+   * @param rightIdentifier  Right Stage Name
+   * @param rightReplacement Right Stage replacement alias
+   */
+  @VisibleForTesting
+  protected void appendOnExpressionClause(String expression,
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/builder/BigQuerySQLBuilder.java]: this seems overly complex, it's hard to follow what is happening, examples would help. In what situations do we need to do this replacement vs just directly using the expression that the user provided? Is it when an invalid table name is being used?

It seems more straightforward and understandable to always use the user provided input as-is and fail if it happens to be invalid (assuming the query gets logged)

Comment (albertshau) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/builder/BigQuerySQLBuilder.java]: this is also the behavior in the Spark implementation, it just uses the user given input as-is.

Comment (fernst) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/builder/BigQuerySQLBuilder.java]: My idea with the re-write was to ensure no conditions can cause a syntax error in BQ (as Spark and BQ have slightly different syntax rules).

I agree that, for simplicity, it would be easier to just paste the input as-is, but I wouldn't like for the user to run into a condition where preview works but then BQ fails.

File: src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/builder/BigQuerySQLBuilder.java (hunk 7 above)
Comment (fernst) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/builder/BigQuerySQLBuilder.java]: Will fix.

File: src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/builder/BigQuerySQLBuilder.java (hunk 6 above)
Comment (fernst) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/builder/BigQuerySQLBuilder.java]: Will fix.

File: src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQuerySQLEngine.java (hunk 5 above)
Comment (fernst) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQuerySQLEngine.java]: Will fix.

File: src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQueryJoinDataset.java (hunk 2 above)
Comment (fernst) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQueryJoinDataset.java]: Good point, let me rename this.

File: src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/builder/BigQuerySQLBuilder.java (hunk 8 above)
Comment (fernst) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/builder/BigQuerySQLBuilder.java]: Oof, good catch.

File: src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQuerySQLEngine.java (hunk 1 above)
Comment (fernst) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQuerySQLEngine.java]: Oh, good catch, I must have deleted the block where I print out a warning. Let me add it again.

File: src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQuerySQLEngine.java (hunk 4 above)
Comment (fernst) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQuerySQLEngine.java]: Take a look now. If issues are found for this stage, I will print one message that mentions the stage and then all error messages concatenated.

File: src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/builder/BigQuerySQLBuilder.java (hunk 10 above)
Comment (fernst) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/builder/BigQuerySQLBuilder.java]: But, to your point, SparkSQL will be as picky as BQ is will something like a non-ascii dataset/table name..

File: src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/builder/BigQuerySQLBuilder.java (hunk 8 above)
Comment (fernst) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/builder/BigQuerySQLBuilder.java]: Will add further regex tests

File: src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/builder/BigQuerySQLBuilder.java (hunk 10 above)
Comment (fernst) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/builder/BigQuerySQLBuilder.java]: After some further testing, I realized your point is pretty valid. Since I'm already checking that stage names and stage aliases do not contain the escape characters for spark ( ` ) and BigQuery ( \ ), supplying the raw SQL without modifications should not cause any problems in BigQuery that were not present in Spark.

File: src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/builder/BigQuerySQLBuilder.java (hunk 8 above)
Comment (fernst) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/builder/BigQuerySQLBuilder.java]: I have removed the regex based on Albert's comments.

File: src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQuerySQLEngine.java (hunk 3 above)
Comment (fernst) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQuerySQLEngine.java]: If you look now, you'll see the issues are used to build a Logger Warning message, which includes the stage name.

Comment (fernst) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQuerySQLEngine.java]: I will add logic to list all invalid fields.

**Review by albertshau — APPROVED** [review]

**Review by tivv — APPROVED** [review]
